### PR TITLE
feat: rescale images to comply with discord size limit

### LINKED ISSUE #199: Add re-scaling settings for screenshots
A user in the Runelite discord reports being unable to get screenshots with an old discord notification plugin due to using an ultra-wide 4K monitor and Discord upload restrictions.

This issue would likely apply to our plugin as well, would be useful if we could offer a setting to scale down the image size, to bypass said restrictions,

### PR DESCRIPTION
**Pull request by iProdigy**
Closes #199
Related https://github.com/MidgetJake/UniversalDiscordNotifier/issues/35

### CODE REVIEW

File: src/main/java/dinkplugin/util/Utils.java [hunk 1]
```diff
@@ -87,6 +89,18 @@ public String getChatBadge(@NotNull AccountType type) {
         }
     }
 
+    public BufferedImage rescale(BufferedImage input, double percent) {
+        if (percent + Math.ulp(1.0) >= 1.0)
+            return input;
+
+        AffineTransform rescale = AffineTransform.getScaleInstance(percent, percent);
+        AffineTransformOp operation = new AffineTransformOp(rescale, AffineTransformOp.TYPE_BILINEAR);
```
Comment (iProdigy) [inline src/main/java/dinkplugin/util/Utils.java:97]: Other options include nearest-neighbor (low quality) and bicubic (slow). Bilinear strikes a good performance-quality balance, and this approach is faster than [Image#getScaledInstance](https://docs.oracle.com/javase/8/docs/api/java/awt/Image.html#getScaledInstance(int,%20int,%20int)) (which has other [problems](https://web.archive.org/web/20200805115809/https://community.oracle.com/docs/DOC-983611), but is still used by [RuneLite](https://github.com/runelite/runelite/blob/ed8076f4d18efe90f59aeedc4c1ef412d9f65b7b/runelite-client/src/main/java/net/runelite/client/util/ImageUtil.java#L259))

File: src/main/java/dinkplugin/message/DiscordMessageHandler.java [hunk 2]
```diff
@@ -162,6 +167,36 @@ private void sendMessage(HttpUrl url, MultipartBody.Builder requestBody, int att
...
+                } catch (IOException e) {
+                    throw new CompletionException("Could not convert image to byte array", e);
+                }
+            })
+            .thenApply(bytes -> {
+                int n = bytes.length;
+                if (n <= Embed.MAX_IMAGE_SIZE)
+                    return bytes; // already compliant; no further rescale necessary
+
+                // calculate scale factor to comply with MAX_IMAGE_SIZE
```
Comment (iProdigy) [inline src/main/java/dinkplugin/message/DiscordMessageHandler.java:187]: With the automatic rescale logic, should we still bother exposing an advanced setting for a specific rescale % ?

**Comment by Felanbird** [comment]
### To test this PR the following needs to be done:


<details>
<summary>Quick Link to skip steps 1-3</summary>

Download the jar file from this PR here: https://github.com/pajlads/DinkPlugin/suites/11849619369/artifacts/619291657

If you download with this link, skip to step 4

</details>

<details>
<summary>Steps 1-3</summary>

1. Click `Checks` at the top of the page

image

2. Click `Build with gradle` on the left side-bar 

image


3. Click `dinkplugin-shadow` which will download the required jar file

image

</details>

4. Un-zip the downloaded file and place `dinkplugin-1.3.2-all.jar` anywhere on your PC

5. Open `Command Prompt`

<details>
<summary>Screenshot</summary>

image

</details>

6. Enter the following into Command Prompt:
`java -ea -jar`

<details>
<summary>Screenshot</summary>

image

</details>

7. Drag & drop the `dinkplugin-1.3.2-all.jar` file into Command Prompt, which will fill in the exact location of the file

<details>
<summary>Screenshot</summary>

image

</details>

8. Hit Enter

<details>
<summary>Screenshot</summary>

image

</details>

File: src/main/java/dinkplugin/util/Utils.java (hunk 1 above)
Comment (Felanbird) [reply src/main/java/dinkplugin/util/Utils.java]: This choice was decided by the fact that the first attempt literally made the sizes larger 🙃

**Review by Felanbird — APPROVED** [review]
I believe this commit works even better than expected.

To test I had a friend with a `3440x1440` monitor boot this PR up and when the image was very obviously above 8MB, discord received an extremely small jpg instead.

6MB png
image

701KB jpg
image

The zoom is slightly different to compare images, as we had to play around with that to get screenshots larger than the 8MB limit

File: src/main/java/dinkplugin/message/DiscordMessageHandler.java [hunk 3]
```diff
@@ -162,6 +170,49 @@ private void sendMessage(HttpUrl url, MultipartBody.Builder requestBody, int att
...
+     * @implNote the image format is either "png" (lossless) or "jpeg" (lossy), both of which can be used in MIME type
+     */
+    private static CompletableFuture<Map.Entry<String, byte[]>> captureScreenshot(DrawManager drawManager, double scalePercent) {
+        CompletableFuture<Image> future = new CompletableFuture<>();
+        drawManager.requestNextFrameListener(future::complete);
+        return future.thenApply(ImageUtil::bufferedImageFromImage)
+            .thenApply(input -> Utils.rescale(input, scalePercent))
+            .thenApply(image -> {
+                try {
+                    String format = "png"; // lossless
```
Comment (iProdigy) [inline src/main/java/dinkplugin/message/DiscordMessageHandler.java:190]: Should there be a config setting to always use lossy over lossless (would typically imply no rescaling necessary)

File: src/main/java/dinkplugin/message/Embed.java [hunk 4]
```diff
@@ -18,6 +18,7 @@
 @Value
 @Builder
 public class Embed {
+    public static final int MAX_IMAGE_SIZE = 8_000_000; // 8 MB
```
Comment (pajlada) [inline src/main/java/dinkplugin/message/Embed.java:21]:
```suggestion
    // The max size of the image before we rescale it to fit Discords file upload limits https://discord.com/developers/docs/reference#uploading-files
    public static final int MAX_IMAGE_SIZE = 8_000_000; // 8MB
```

Technically I think the upload limit was very recently upped to 25MB, however I think this is still a reasonable change to make, as we don't want to hog the network to upload a 24MB 8K PNG imo. A separate PR can make either this number configurable (as it depends on the boost level of the server) or some setting to just disable scaling as a whole.

File: src/main/java/dinkplugin/message/DiscordMessageHandler.java [hunk 5]
```diff
@@ -162,6 +170,49 @@ private void sendMessage(HttpUrl url, MultipartBody.Builder requestBody, int att
...
+                // calculate scale factor to comply with MAX_IMAGE_SIZE
+                double factor = Math.sqrt(1.0 * Embed.MAX_IMAGE_SIZE / n);
+
+                // bytes => original image => rescaled image => updated bytes
+                try (InputStream is = new ByteArrayInputStream(bytes)) {
+                    String format = "jpeg"; // lossy
+                    BufferedImage rescaled = Utils.rescale(ImageIO.read(is), factor);
+                    return Pair.of(format, Utils.convertImageToByteArray(rescaled, format));
+                } catch (Exception e) {
+                    throw new CompletionException("Failed to automatically resize image below Discord size limit", e);
```
Comment (pajlada) [inline src/main/java/dinkplugin/message/DiscordMessageHandler.java:211]:
```suggestion
                    throw new CompletionException("Failed to resize image below Discord size limit", e);
```

nitpick: the word `automatically` here doesn't provide anything useful

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,7 @@
 ## Unreleased
 
+- Bugfix: Rescale screenshots to comply with Discord's 8MB size limit. (#200)
```
Comment (pajlada) [inline CHANGELOG.md:3]: nitpick: I don't think this is a bug, I'd call it a minor feature instead

File: README.md [hunk 7]
```diff
@@ -39,6 +39,7 @@ Some notifiers require in-game settings to be configured to send chat messages u
 * Multiple webhook urls are supported; simply place each on a separate line
 * Each notifier can send webhook messages to separate "override" urls
 * Screenshots can be individually configured for each notifier
+* Screenshot files are automatically compressed if needed to comply with Discord limits
```
Comment (pajlada) [inline README.md:42]:
```suggestion
* Screenshots are compressed if needed to comply with Discord limits
```

**Review by pajlada — CHANGES_REQUESTED** [review]
Just nitpicks, looks good to me!

**Review by pajlada — APPROVED** [review]
Looks good to me! :baguette_bread:

**Pull request merged**